rasa/shared/core/domain.py

```diff
@@ -1849,10 +1838,32 @@ def _validate_slot_mappings(forms: Union[Dict, List]) -> None:
     if not isinstance(forms, dict):
         raise InvalidDomain("Forms have to be specified as dictionary.")
 
-    for form_name, slots in forms.items():
-        if slots is None:
+    for form_name, form_data in forms.items():
+        if form_data is None:
             continue
 
+        if isinstance(form_data, Dict):
+            if "required_slots" in form_data:
+                slots = forms[form_name].get("required_slots")
+            elif "global_not_intent" in form_data and "required_slots" not in form_data:
+                raise InvalidDomain(
+                    f"If you use the `global_not_intent` parameter in your form, then"
+                    f" the keyword `required_slots` should precede the definition"
+                    f" of your slot mappings. Please see"
+                    f" {rasa.shared.constants.DOCS_URL_FORMS} for more information."
+                )
+            else:
+                slots = form_data
+
+        if "required_slots" not in form_data:
+            rasa.shared.utils.io.raise_deprecation_warning(
+                f"The definition of slot mappings in your form"
+                f" should always be preceded by the keyword `required_slots`."
+                f" The lack of this keyword will be deprecated in"
+                f" Rasa Open Source 3.0.0. Please see "
+                f"{rasa.shared.constants.DOCS_URL_FORMS} for more information.",
+            )
+
         if not isinstance(slots, Dict):
```

Comment:
pycharm gives me a warning that `slots` might be referenced before assignment. We should fix this.